### PR TITLE
Refactor build, add first pass service-worker-generation

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: src/build/build.ts [hunk 1]
```diff
@@ -59,38 +73,51 @@ export function build(entrypoint, sources): Promise<any> {
...
+      .pipe(vfs.dest('build/unbundled'))
+
+    let bundledPhase = new ForkedVinylStream(allFiles)
+      .pipe(bundler.bundle)
+      .pipe(vfs.dest('build/bundled'));
+
+    waitForAll([unbundledPhase, bundledPhase])
+      .then((_) => {
+        console.log('all done!');
+        let deps = Array.from(bundler.streamResolver.requestedUrls.values());
```
Comment (azakus) [inline src/build/build.ts]: TODO: abstract over this

File: src/build/bundle.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  constructor(root: string, shell: string, entrypoints: string[]) {
+    this.root = root;
+    this.shell = shell;
+    let allEntrypoints = [];
+    if (shell) allEntrypoints.push(shell);
+    this.allEntrypoints =
+        Array.prototype.concat.apply(allEntrypoints, entrypoints);
+
+    console.log('allEntrypoints', this.allEntrypoints);
```
Comment (azakus) [inline src/build/bundle.ts]: Remove

File: src/build/bundle.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      inlineScripts: true,
+      inlineCss: true,
+      stripComments: true,
+    });
+
+    this.bundle = compose([
+      this.streamResolver,
+      gulpif((file) => this.isEntryPoint(file.path),
+        compose([
+          new Logger('pre-vulcanize'),
```
Comment (azakus) [inline src/build/bundle.ts]: Remove Logger streams

File: src/build/merge-stream.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import {PassThrough, Stream} from 'stream';
```
Comment (azakus) [inline src/build/merge-stream.ts]: This file still necessary?

File: src/build/streams.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    });
+    input.on('end', () => {
+      this.emit('end');
+    });
+    input.on('error', (e) => {
+      this.emit('error', e);
+    });
+  }
+
+  _read(size: number) {
```
Comment (azakus) [inline src/build/streams.ts]: Let's comment why this is necessary at all (downstream pipe will poll this to eagerly read, link to node docs?).

File: src/build/sw-precache.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import * as fs from 'fs';
+import * as path from 'path';
+
+// non-ES compatible modules
+const swPrecache = require('sw-precache');
+
+export function generateServiceWorker(
```
Comment (azakus) [inline src/build/sw-precache.ts]: Comment on the hardcoded values
TODO: merge in a config to supply these values

File: src/build/uglify-transform.ts [hunk 7]
```diff
@@ -28,7 +28,7 @@ export class UglifyTransform extends Transform {
         file.contents = new Buffer(contents);
       } catch (err) {
         console.error('Could not uglify',file.path);
-        console.error(err.stack);
+        // console.error(err.stack);
```
Comment (azakus) [inline src/build/uglify-transform.ts]: remove?

File: src/build/vulcanize.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  opts: any;
+  count: number = 0;
+
+  constructor(opts) {
+    super({objectMode: true});
+    this.opts = opts;
+  }
+
+  _transform(file: File, enc: string, cb: (error?, data?) => void): void {
+    if (this.count++ > 0) {
```
Comment (azakus) [inline src/build/vulcanize.ts]: Is this still necessary?

File: src/build/vulcanize.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+      cb(null, file);
+      return;
+    }
+
+    if (file.isNull()) {
+      cb(null, file);
+			return;
```
Comment (azakus) [inline src/build/vulcanize.ts]: Formatting?

File: test/build/optimize_test.js [hunk 10]
```diff
@@ -16,19 +16,21 @@ const PassThrough = require('stream').PassThrough;
...
     });
     stream.on('error', (err) => cb(err));
   }
+
   test('no options', () => {
     let stream = new PassThrough();
-    let op = optimizePipe(stream);
-    assert.equal(stream, op, 'stream should be identical if no options given');
+    let op = stream.pipe(optimize());
+    // assert.equal(stream, op, 'stream should be identical if no options given');
```
Comment (azakus) [inline test/build/optimize_test.js]: this now has no assertion tests, maybe just remove?

File: src/build/build.ts (hunk 1 above)
Comment (justinfagnani) [inline src/build/build.ts]: done

File: src/build/bundle.ts (hunk 2 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: done

File: src/build/bundle.ts (hunk 3 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: done

File: src/build/merge-stream.ts (hunk 4 above)
Comment (justinfagnani) [inline src/build/merge-stream.ts]: removed

File: src/build/streams.ts (hunk 5 above)
Comment (justinfagnani) [inline src/build/streams.ts]: done

File: src/build/sw-precache.ts (hunk 6 above)
Comment (justinfagnani) [inline src/build/sw-precache.ts]: done

File: src/build/uglify-transform.ts (hunk 7 above)
Comment (justinfagnani) [inline src/build/uglify-transform.ts]: done

File: src/build/vulcanize.ts (hunk 8 above)
Comment (justinfagnani) [inline src/build/vulcanize.ts]: done

File: src/build/vulcanize.ts (hunk 9 above)
Comment (justinfagnani) [inline src/build/vulcanize.ts]: weird...

File: test/build/optimize_test.js (hunk 10 above)
Comment (justinfagnani) [inline test/build/optimize_test.js]: done

**Comment by justinfagnani** [comment]
@azakus PTAL

**Comment by azakus** [comment]
LGTM

**Pull request merged**